api/mfx_dispatch/windows/include/mfx_dispatcher_log.h

```diff
@@ -0,0 +1,282 @@
+#define ZERO_OR_SPACE(value) ((0==(value)) ? '0' : (value))
+#define MFXU32TOFOURCC(mfxu32)\
+    ZERO_OR_SPACE((char)(mfxu32 & 0xFF)), \
+    ZERO_OR_SPACE((char)((mfxu32 >> 8) & 0xFF)),\
+    ZERO_OR_SPACE((char)((mfxu32 >> 16) & 0xFF)),\
+    ZERO_OR_SPACE((char)((mfxu32 >> 24) & 0xFF))
+
+#define MFXGUIDTYPE() "%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X-%X"
+
+#define MFXGUIDTOHEX(guid)\
+    (guid)->Data[0],\
+    (guid)->Data[1],\
+    (guid)->Data[2],\
+    (guid)->Data[3],\
+    (guid)->Data[4],\
+    (guid)->Data[5],\
+    (guid)->Data[6],\
+    (guid)->Data[7],\
+    (guid)->Data[8],\
+    (guid)->Data[9],\
+    (guid)->Data[10],\
+    (guid)->Data[11],\
+    (guid)->Data[12],\
+    (guid)->Data[13],\
+    (guid)->Data[14],\
+    (guid)->Data[15]
+
+#if defined(MFX_DISPATCHER_LOG)
+
+//---------------------------setup section------------------------
+//using of formating instead of variadic macro with NULL end, 
+//leads to more flexibility in format, however constructing string 
+//with vsprintf_s is a time wasting
+#define DISPATCHER_LOG_USE_FORMATING 1
+
+//creates unique object, event guid registration, factories on heap
+//heap reduce stack allocation and reduce reservation time at startup
+//is a vital if mediasdk wont use
+#define DISPATCHER_LOG_HEAP_SINGLETONES
+
+// guid for all dispatcher events
+#define DISPATCHER_LOG_EVENT_GUID L"{EB0538CC-4FEE-484d-ACEE-1182E9F37A57}"
+
+//puts a sink into listeners list
+//#define DISPATCHER_LOG_REGISTER_EVENT_PROVIDER
+
+//puts a sink into listeners list
+//#define DISPATCHER_LOG_REGISTER_FILE_WRITER
+#define DISPACTHER_LOG_FW_PATH "c:\\dispatcher.log"
+
+#include <stdio.h>
+#include <stdarg.h>
+
+//callback interface for intercept logging messages
+class IMsgHandler
+{
+public:
+    virtual ~IMsgHandler(){}
+    virtual void Write(int level, int opcode, const char * msg, va_list argptr) = 0;
+};
+
+#if  DISPATCHER_LOG_USE_FORMATING
+
+    #define DISPATCHER_LOG(lvl, opcode, str)\
+    {\
+        DispatcherLogBracketsHelper wrt(lvl,opcode);\
+        wrt.Write str;\
+    }
+#else
+    #define DISPATCHER_LOG_VA_ARGS(...) wrt.Write(__VA_ARGS__, NULL)
+    //WARNING: don't use types that occupy more that 4 bytes in memory 
+    //WARNING: don't use %s in format specifier
+    #define DISPATCHER_LOG(lvl, opcode, str) \
+    {\
+        DispatcherLogBracketsHelper wrt(lvl, opcode);\
+        DISPATCHER_LOG_VA_ARGS str;\
+    }
+#endif//DISPATCHER_LOG_USE_FORMATING
+
+#define DISPATCHER_LOG_OPERATION(operation) operation
+
+#define __name_from_line( name, line ) name ## line
+#define _name_from_line( name , line) __name_from_line( name, line ) 
+#define name_from_line( name ) _name_from_line( name, __LINE__) 
+
+
+#define DISPATCHER_LOG_AUTO(lvl, msg)\
+    DispatchLogBlockHelper name_from_line(__auto_log_)(lvl); name_from_line(__auto_log_).Write msg;
+
+#include <memory>
+#include <map>
+#include <list>
+#include <string>
+
+template <class T> 
+class DSSingleTone
```

Comment:
`SingleTone` means single tone? What is this? I believe that pattern [Singleton](https://en.wikipedia.org/wiki/Singleton_pattern) was implied.